File: python/src/dual_agent_sdk/convergence.py
"""6-layer convergence protocol (Chain of Responsibility).

Implements :class:`ConvergenceEngine`, which chains six ordered checks to
determine whether the Generator-Critic loop should accept the current
artifact, continue iterating, or escalate a deadlock.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from math import isclose
from typing import Optional

from .loop_detector import SemanticLoopDetector
from .models import AgentConfig, Decision, RoundRecord, Verdict


# ======================================================================
# Individual convergence checks
# ======================================================================

class ConvergenceCheck(ABC):
    """Abstract base for a single convergence rule.

    Each concrete check is a link in the Chain of Responsibility.  It
    receives the full round context and returns a :class:`Decision` or
    ``None`` to pass control to the next check.
    """

    @abstractmethod
    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        """Evaluate this check.

        Returns:
            A decision if this check resolves the round, or ``None`` to
            fall through to the next check in the chain.
        """
        ...


class HardCeilingCheck(ConvergenceCheck):
    """**Layer 1** -- Hard round cap.

    If the current round number has reached *max_rounds* the loop must
    terminate, returning ``"accept"`` (the best artifact so far).
    """

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        if round_num >= config.max_rounds:
            return "accept"
        return None


class QualityThresholdCheck(ConvergenceCheck):
    """**Layer 2** -- Quality gate.

    Accepts immediately when *verdict.score* >= *config.quality_threshold*
    AND there are no blocking issues.
    """

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        if verdict.score >= config.quality_threshold and not verdict.is_blocking:
            return "accept"
        return None


class ScoreConvergenceCheck(ConvergenceCheck):
    """**Layer 3** -- Score delta.

    Accepts when the absolute score change between successive rounds is
    smaller than *config.convergence_threshold* (diminishing returns).
    """

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        if prev_verdict is None:
            return None
        if isclose(verdict.score, prev_verdict.score, abs_tol=config.convergence_threshold):
            return "accept"
        return None


class IssueDecayCheck(ConvergenceCheck):
    """**Layer 4** -- Severity-based issue filter.

    Gradually narrows the set of actionable issue severities so that
    later rounds only address critical problems:

    * Round 1  → *critical* + *major*
    * Round 2  → *critical* only
    * Round 3+ → no issues are actionable → accept

    When there are no issues at all, this check passes through (None)
    so other layers (e.g. SemanticLoopCheck) can still evaluate.
    """

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        # No issues at all → let other layers decide
        if not verdict.issues:
            return None

        if round_num == 1:
            allowed = {"critical", "major"}
        elif round_num == 2:
            allowed = {"critical"}
        else:
            # Round 3+: no issue severity is actionable
            return "accept"

        actionable = [i for i in verdict.issues if i.severity in allowed]
        if not actionable:
            return "accept"
        return None


class ROICheck(ConvergenceCheck):
    """**Layer 5** -- Return-on-investment / diminishing returns.

    Accepts when the improvement from the previous round to the current
    round is less than ``0.05 * (roi_decay_factor ** (round_num - 1))``.
    This means the bar for "meaningful improvement" rises exponentially
    across rounds, preventing marginal gains from extending the loop.

    Uses rounding to avoid floating-point precision issues at boundary values.
    """

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        if prev_verdict is None:
            return None
        # Round to 10 decimal places to avoid floating-point noise
        improvement = round(verdict.score - prev_verdict.score, 10)
        min_improvement = round(0.05 * (config.roi_decay_factor ** (round_num - 1)), 10)
        if improvement <= min_improvement:
            return "accept"
        return None


class SemanticLoopCheck(ConvergenceCheck):
    """**Layer 6** -- Semantic repetition detection.

    Uses :class:`SemanticLoopDetector` to detect if the generator is
    producing near-identical output across rounds.  When a loop is
    detected the result is ``"escalate"`` (if enabled in config) so a
    meta-judge can break the tie.
    """

    def __init__(self) -> None:
        super().__init__()
        self._detector: Optional[SemanticLoopDetector] = None

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision | None:
        # Need at least two rounds to detect a semantic loop.
        if len(history) < 2:
            return None

        # Create a fresh detector per evaluation to avoid
        # accumulating duplicate seeds across repeated calls.
        detector = SemanticLoopDetector(
            threshold=config.loop_similarity_threshold,
            window_size=3,
        )

        # Seed the detector with all prior rounds (not the current one).
        for prior in history[:-1]:
            detector.add(prior.artifact.content)

        # Check the current round against the window of prior rounds.
        current_content = history[-1].artifact.content
        if detector.is_looping(current_content):
            return "escalate" if config.enable_escalation else "accept"

        return None


# ======================================================================
# Chain-of-Responsibility engine
# ======================================================================

class ConvergenceEngine:
    """Orchestrates the six convergence checks in priority order.

    Usage::

        engine = ConvergenceEngine()
        decision = engine.evaluate(round_num, verdict, prev_verdict,
                                   config, history)
    """

    def __init__(self) -> None:
        self._checks: list[ConvergenceCheck] = [
            HardCeilingCheck(),
            QualityThresholdCheck(),
            ScoreConvergenceCheck(),
            IssueDecayCheck(),
            ROICheck(),
            SemanticLoopCheck(),
        ]

    def evaluate(
        self,
        round_num: int,
        verdict: Verdict,
        prev_verdict: Verdict | None,
        config: AgentConfig,
        history: list[RoundRecord],
    ) -> Decision:
        """Run the chain and return the first non-``None`` decision.

        If every check returns ``None`` the engine defaults to
        ``"continue"``, meaning another round is warranted.
        """
        for check in self._checks:
            decision = check.evaluate(round_num, verdict, prev_verdict, config, history)
            if decision is not None:
                return decision
        return "continue"

File: typescript/src/convergence.ts
import { Verdict, RoundRecord, AgentConfig, Decision } from './models.js';
import { SemanticLoopDetector } from './loop-detector.js';

// ---------------------------------------------------------------------------
// Chain-of-Responsibility convergence checks
// Each check returns a Decision or null to pass control to the next check.
// ---------------------------------------------------------------------------

/**
 * Abstract base class for a single convergence check.
 * Subclasses implement {@link evaluate} and return a {@link Decision}
 * or null to hand off to the next check in the chain.
 */
abstract class ConvergenceCheck {
  /**
   * Evaluate this check's condition.
   *
   * @param roundNum - Current round number (1-based).
   * @param verdict - The Critic's verdict for the current round.
   * @param prevVerdict - The Critic's verdict from the previous round (null on round 1).
   * @param config - The active agent configuration.
   * @param history - All round records accumulated so far.
   * @returns A Decision if this check triggers, or null to continue the chain.
   */
  abstract evaluate(
    roundNum: number,
    verdict: Verdict,
    prevVerdict: Verdict | null,
    config: AgentConfig,
    history: RoundRecord[],
  ): Decision | null;
}

// ---------------------------------------------------------------------------
// Layer 1: Hard ceiling
// ---------------------------------------------------------------------------

/**
 * Forces acceptance when the maximum number of rounds is reached.
 * This is the safety valve that guarantees termination.
 */
class HardCeilingCheck extends ConvergenceCheck {
  evaluate(
    roundNum: number,
    _verdict: Verdict,
    _prevVerdict: Verdict | null,
    config: AgentConfig,
    _history: RoundRecord[],
  ): Decision | null {
    if (roundNum >= config.maxRounds) {
      return 'accept';
    }
    return null;
  }
}

// ---------------------------------------------------------------------------
// Layer 2: Quality threshold
// ---------------------------------------------------------------------------

/**
 * Accepts the artifact when its score meets or exceeds the quality threshold
 * AND there are no blocking issues.
 */
class QualityThresholdCheck extends ConvergenceCheck {
  evaluate(
    _roundNum: number,
    verdict: Verdict,
    _prevVerdict: Verdict | null,
    config: AgentConfig,
    _history: RoundRecord[],
  ): Decision | null {
    if (verdict.score >= config.qualityThreshold && !verdict.isBlocking) {
      return 'accept';
    }
    return null;
  }
}

// ---------------------------------------------------------------------------
// Layer 3: Score convergence
// ---------------------------------------------------------------------------

/**
 * Accepts when the score delta between the current and previous round
 * falls below the convergence threshold, indicating diminishing returns.
 */
class ScoreConvergenceCheck extends ConvergenceCheck {
  evaluate(
    _roundNum: number,
    verdict: Verdict,
    prevVerdict: Verdict | null,
    config: AgentConfig,
    _history: RoundRecord[],
  ): Decision | null {
    if (prevVerdict === null) return null;
    if (Math.abs(verdict.score - prevVerdict.score) < config.convergenceThreshold) {
      return 'accept';
    }
    return null;
  }
}

// ---------------------------------------------------------------------------
// Layer 4: Issue decay
// ---------------------------------------------------------------------------

/**
 * Applies a progressive filter to issues across rounds.
 * Round 1: critical + major issues remain actionable.
 * Round 2: only critical issues remain actionable.
 * Round 3+: no issues are considered actionable — accept.
 */
class IssueDecayCheck extends ConvergenceCheck {
  evaluate(
    roundNum: number,
    verdict: Verdict,
    _prevVerdict: Verdict | null,
    _config: AgentConfig,
    _history: RoundRecord[],
  ): Decision | null {
    const remaining = this.filterActionableIssues(verdict.issues, roundNum);
    if (remaining.length === 0) {
      return 'accept';
    }
    return null;
  }

  /**
   * Determine which issues are still actionable at a given round.
   * Round 1: keep critical + major.
   * Round 2: keep critical only.
   * Round 3+: nothing is actionable.
   */
  private filterActionableIssues(
    issues: { severity: string }[],
    roundNum: number,
  ): { severity: string }[] {
    if (roundNum >= 3) return [];
    if (roundNum === 2) return issues.filter((i) => i.severity === 'critical');
    // Round 1
    return issues.filter((i) => i.severity === 'critical' || i.severity === 'major');
  }
}

// ---------------------------------------------------------------------------
// Layer 5: ROI (return-on-investment) check
// ---------------------------------------------------------------------------

/**
 * Accepts when the improvement between rounds is smaller than
 * a decaying threshold: 0.05 * (roiDecayFactor^(roundNum-1)).
 *
 * This models diminishing returns — each additional round is expected
 * to yield exponentially less improvement.
 */
class ROICheck extends ConvergenceCheck {
  evaluate(
    roundNum: number,
    verdict: Verdict,
    prevVerdict: Verdict | null,
    config: AgentConfig,
    _history: RoundRecord[],
  ): Decision | null {
    if (prevVerdict === null) return null;
    const improvement = verdict.score - prevVerdict.score;
    const threshold = 0.05 * Math.pow(config.roiDecayFactor, roundNum - 1);
    if (improvement < threshold) {
      return 'accept';
    }
    return null;
  }
}

// ---------------------------------------------------------------------------
// Layer 6: Semantic loop detection
// ---------------------------------------------------------------------------

/**
 * Detects semantic loops by delegating to the {@link SemanticLoopDetector}.
 * When a loop is detected and escalation is enabled, returns 'escalate'.
 * When a loop is detected and escalation is disabled, returns 'accept'.
 */
class SemanticLoopCheck extends ConvergenceCheck {
  private loopDetector: SemanticLoopDetector;

  constructor(loopDetector: SemanticLoopDetector) {
    super();
    this.loopDetector = loopDetector;
  }

  evaluate(
    _roundNum: number,
    verdict: Verdict,
    _prevVerdict: Verdict | null,
    config: AgentConfig,
    _history: RoundRecord[],
  ): Decision | null {
    // Feed the verdict's suggestion (or a combo of content) into the loop detector.
    // We primarily check the suggestion text, which captures what the Critic wants changed.
    const checkText =
      verdict.suggestion ??
      verdict.issues.map((i) => i.description).join(' ') ??
      '';

    if (checkText.length === 0) return null;

    if (this.loopDetector.isLooping(checkText)) {
      return config.enableEscalation ? 'escalate' : 'accept';
    }

    // Also register this round's text so future checks can compare against it.
    this.loopDetector.add(checkText);
    return null;
  }
}

// ---------------------------------------------------------------------------
// Convergence Engine
// ---------------------------------------------------------------------------

/**
 * The ConvergenceEngine orchestrates the 6-layer chain-of-responsibility
 * convergence protocol. Each layer is checked in order; the first layer
 * that returns a non-null {@link Decision} short-circuits the chain.
 *
 * Layers:
 * 1. Hard ceiling — forced accept at maxRounds.
 * 2. Quality threshold — accept if score meets threshold and no blockers.
 * 3. Score convergence — accept if score delta is tiny.
 * 4. Issue decay — accept if no actionable issues remain by round.
 * 5. ROI check — accept if improvement is below diminishing-returns threshold.
 * 6. Semantic loop — escalate to meta-judge if the agents are repeating.
 */
export class ConvergenceEngine {
  private checks: ConvergenceCheck[];
  private loopDetector: SemanticLoopDetector;

  constructor(config: AgentConfig) {
    this.loopDetector = new SemanticLoopDetector(config.loopSimilarityThreshold);
    this.checks = [
      new HardCeilingCheck(),
      new QualityThresholdCheck(),
      new ScoreConvergenceCheck(),
      new IssueDecayCheck(),
      new ROICheck(),
      new SemanticLoopCheck(this.loopDetector),
    ];
  }

  /**
   * Evaluate all convergence checks in priority order.
   *
   * @param roundNum - Current round number (1-based).
   * @param verdict - The Critic's verdict for this round.
   * @param prevVerdict - The Critic's verdict from the previous round (null on round 1).
   * @param config - The active agent configuration.
   * @param history - All round records accumulated so far.
   * @returns The first non-null decision, or 'continue' if no check triggers.
   */
  evaluate(
    roundNum: number,
    verdict: Verdict,
    prevVerdict: Verdict | null,
    config: AgentConfig,
    history: RoundRecord[],
  ): Decision {
    for (const check of this.checks) {
      const decision = check.evaluate(roundNum, verdict, prevVerdict, config, history);
      if (decision !== null) {
        return decision;
      }
    }
    return 'continue';
  }

  /**
   * Expose the underlying loop detector for external inspection or reset.
   */
  get detector(): SemanticLoopDetector {
    return this.loopDetector;
  }
}
